### ISSUE TITLE
RTMDet-ins low AP/AR for "coco small" objects

### ISSUE THREAD

**Issue by roman-duris** [open-mmlab/mmdetection#10874]
I have been training RTMDet-ins model for custom dataset with 1 class.
Overall precision is pretty good, however the models even L and X configurations have problems with small objects. 
![image](https://github.com/open-mmlab/mmdetection/assets/141549520/45d1d6d8-64df-4180-b1d3-f03d999c2ed4)

I have been trying to improve segmentation of smaller objects. 
I tried to increase depth of convolutions in Head, and backbone, increasing input and output channels throughout the network as well as playing with pred_kernel_size parameter in the head. Sadly all those tests did not result in significant increase of AR/AP for small objects. 

Therefore I would be very happy if you had any suggestions on what to try out to improve the detection of small objects.

My configuration file looks like this: 

```
default_scope = 'mmdet'
default_hooks = dict(
    timer=dict(type='IterTimerHook'),
    logger=dict(type='LoggerHook', interval=50),
    param_scheduler=dict(type='ParamSchedulerHook'),
    checkpoint=dict(type='CheckpointHook', interval=10, max_keep_ckpts=3),
    sampler_seed=dict(type='DistSamplerSeedHook'),
    visualization=dict(type='DetVisualizationHook'))
env_cfg = dict(
    cudnn_benchmark=False,
    mp_cfg=dict(mp_start_method='fork', opencv_num_threads=0),
    dist_cfg=dict(backend='nccl'))
vis_backends = [
    dict(type='LocalVisBackend'),
]
visualizer = dict(
    type='DetLocalVisualizer',
    vis_backends=[
        dict(type='LocalVisBackend'),
    ],
    name='visualizer')
log_processor = dict(type='LogProcessor', window_size=50, by_epoch=True)
log_level = 'INFO'
load_from = '/workspace/models/rtmdet_ins_l/epoch_300.pth'
resume = False
train_cfg = dict(
    type='EpochBasedTrainLoop',
    max_epochs=300,
    val_interval=10,
    dynamic_intervals=[
        (
            280,
            1,
        ),
    ])
val_cfg = dict(type='ValLoop')
test_cfg = dict(type='TestLoop')
param_scheduler = [
    dict(
        type='LinearLR', start_factor=1e-05, by_epoch=False, begin=0,
        end=1000),
    dict(
        type='CosineAnnealingLR',
        eta_min=0.0002,
        begin=150,
        end=300,
        T_max=150,
        by_epoch=True,
        convert_to_iter_based=True),
]
optim_wrapper = dict(
    type='OptimWrapper',
    optimizer=dict(type='AdamW', lr=0.004, weight_decay=0.05),
    paramwise_cfg=dict(
        norm_decay_mult=0, bias_decay_mult=0, bypass_duplicate=True))
auto_scale_lr = dict(enable=True, base_batch_size=16)
dataset_type = 'CocoDataset'
data_root = '/workspace/datasets/[redacted]/'
backend_args = None
train_pipeline = [
    dict(type='LoadImageFromFile', backend_args=None),
    dict(
        type='LoadAnnotations',
        with_bbox=True,
        with_mask=True,
        poly2mask=False),
    dict(type='CachedMosaic', img_scale=(
        640,
        640,
    ), pad_val=114.0),
    dict(
        type='RandomResize',
        scale=(
            1280,
            1280,
        ),
        ratio_range=(
            0.1,
            2.0,
        ),
        keep_ratio=True),
    dict(
        type='RandomCrop',
        crop_size=(
            640,
            640,
        ),
        recompute_bbox=True,
        allow_negative_crop=True),
    dict(type='YOLOXHSVRandomAug'),
    dict(type='RandomFlip', prob=0.5),
    dict(type='Pad', size=(
        640,
        640,
    ), pad_val=dict(img=(
        114,
        114,
        114,
    ))),
    dict(
        type='CachedMixUp',
        img_scale=(
            640,
            640,
        ),
        ratio_range=(
            1.0,
            1.0,
        ),
        max_cached_images=20,
        pad_val=(
            114,
            114,
            114,
        )),
    dict(type='FilterAnnotations', min_gt_bbox_wh=(
        1,
        1,
    )),
    dict(type='PackDetInputs'),
]
test_pipeline = [
    dict(type='LoadImageFromFile', backend_args=None),
    dict(type='Resize', scale=(
        640,
        640,
    ), keep_ratio=True),
    dict(type='Pad', size=(
        640,
        640,
    ), pad_val=dict(img=(
        114,
        114,
        114,
    ))),
    dict(
        type='PackDetInputs',
        meta_keys=(
            'img_id',
            'img_path',
            'ori_shape',
            'img_shape',
            'scale_factor',
        )),
]
train_dataloader = dict(
    batch_size=8,
    num_workers=1,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=True),
    batch_sampler=None,
    dataset=dict(
        type='CocoDataset',
        data_root='/workspace/datasets/[redacted]/',
        ann_file='annotations/instances_Train.json',
        data_prefix=dict(img='images/'),
        filter_cfg=dict(filter_empty_gt=True, min_size=32),
        pipeline=[
            dict(type='LoadImageFromFile', backend_args=None),
            dict(
                type='LoadAnnotations',
                with_bbox=True,
                with_mask=True,
                poly2mask=False),
            dict(type='CachedMosaic', img_scale=(
                640,
                640,
            ), pad_val=114.0),
            dict(
                type='RandomResize',
                scale=(
                    1280,
                    1280,
                ),
                ratio_range=(
                    0.1,
                    2.0,
                ),
                keep_ratio=True),
            dict(
                type='RandomCrop',
                crop_size=(
                    640,
                    640,
                ),
                recompute_bbox=True,
                allow_negative_crop=True),
            dict(type='YOLOXHSVRandomAug'),
            dict(type='RandomFlip', prob=0.5),
            dict(
                type='Pad',
                size=(
                    640,
                    640,
                ),
                pad_val=dict(img=(
                    114,
                    114,
                    114,
                ))),
            dict(
                type='CachedMixUp',
                img_scale=(
                    640,
                    640,
                ),
                ratio_range=(
                    1.0,
                    1.0,
                ),
                max_cached_images=20,
                pad_val=(
                    114,
                    114,
                    114,
                )),
            dict(type='FilterAnnotations', min_gt_bbox_wh=(
                1,
                1,
            )),
            dict(type='PackDetInputs'),
        ],
        backend_args=None,
        metainfo=dict(classes='[redacted]', palette=[
            (
                220,
                20,
                60,
            ),
        ])),
    pin_memory=True)
val_dataloader = dict(
    batch_size=5,
    num_workers=1,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type='CocoDataset',
        data_root='/workspace/datasets/[redacted]/',
        ann_file='annotations/instances_Validation.json',
        data_prefix=dict(img='images/'),
        test_mode=True,
        pipeline=[
            dict(type='LoadImageFromFile', backend_args=None),
            dict(type='Resize', scale=(
                640,
                640,
            ), keep_ratio=True),
            dict(
                type='Pad',
                size=(
                    640,
                    640,
                ),
                pad_val=dict(img=(
                    114,
                    114,
                    114,
                ))),
            dict(
                type='PackDetInputs',
                meta_keys=(
                    'img_id',
                    'img_path',
                    'ori_shape',
                    'img_shape',
                    'scale_factor',
                )),
        ],
        backend_args=None,
        metainfo=dict(classes='[redacted]', palette=[
            (
                220,
                20,
                60,
            ),
        ])))
test_dataloader = dict(
    batch_size=5,
    num_workers=1,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type='CocoDataset',
        data_root='/workspace/datasets/[redacted]/',
        ann_file='annotations/instances_Test.json',
        data_prefix=dict(img='images/'),
        test_mode=True,
        pipeline=[
            dict(type='LoadImageFromFile', backend_args=None),
            dict(type='Resize', scale=(
                640,
                640,
            ), keep_ratio=True),
            dict(
                type='Pad',
                size=(
                    640,
                    640,
                ),
                pad_val=dict(img=(
                    114,
                    114,
                    114,
                ))),
            dict(
                type='PackDetInputs',
                meta_keys=(
                    'img_id',
                    'img_path',
                    'ori_shape',
                    'img_shape',
                    'scale_factor',
                )),
        ],
        backend_args=None,
        metainfo=dict(classes='[redacted]', palette=[
            (
                220,
                20,
                60,
            ),
        ])))
val_evaluator = dict(
    type='CocoMetric',
    ann_file=
    '/workspace/datasets/[redacted]//annotations/instances_Validation.json',
    metric=[
        'bbox',
        'segm',
    ],
    format_only=False,
    backend_args=None,
    proposal_nums=(
        100,
        1,
        10,
    ))
test_evaluator = dict(
    type='CocoMetric',
    ann_file='/workspace/datasets/[redacted]//annotations/instances_Test.json',
    metric=[
        'bbox',
        'segm',
    ],
    format_only=False,
    backend_args=None,
    proposal_nums=(
        100,
        1,
        10,
    ))
tta_model = dict(
    type='DetTTAModel',
    tta_cfg=dict(nms=dict(type='nms', iou_threshold=0.6), max_per_img=100))
img_scales = [
    (
        640,
        640,
    ),
    (
        320,
        320,
    ),
    (
        960,
        960,
    ),
]
tta_pipeline = [
    dict(type='LoadImageFromFile', backend_args=None),
    dict(
        type='TestTimeAug',
        transforms=[
            [
                dict(type='Resize', scale=(
                    640,
                    640,
                ), keep_ratio=True),
                dict(type='Resize', scale=(
                    320,
                    320,
                ), keep_ratio=True),
                dict(type='Resize', scale=(
                    960,
                    960,
                ), keep_ratio=True),
            ],
            [
                dict(type='RandomFlip', prob=1.0),
                dict(type='RandomFlip', prob=0.0),
            ],
            [
                dict(
                    type='Pad',
                    size=(
                        960,
                        960,
                    ),
                    pad_val=dict(img=(
                        114,
                        114,
                        114,
                    ))),
            ],
            [
                dict(
                    type='PackDetInputs',
                    meta_keys=(
                        'img_id',
                        'img_path',
                        'ori_shape',
                        'img_shape',
                        'scale_factor',
                        'flip',
                        'flip_direction',
                    )),
            ],
        ]),
]
model = dict(
    type='RTMDet',
    data_preprocessor=dict(
        type='DetDataPreprocessor',
        mean=[
            103.53,
            116.28,
            123.675,
        ],
        std=[
            57.375,
            57.12,
            58.395,
        ],
        bgr_to_rgb=False,
        batch_augments=None),
    backbone=dict(
        type='CSPNeXt',
        arch='P5',
        expand_ratio=0.5,
        deepen_factor=1,
        widen_factor=1,
        channel_attention=True,
        norm_cfg=dict(type='SyncBN'),
        act_cfg=dict(type='SiLU', inplace=True)),
    neck=dict(
        type='CSPNeXtPAFPN',
        in_channels=[
            256,
            512,
            1024,
        ],
        out_channels=256,
        num_csp_blocks=3,
        expand_ratio=0.5,
        norm_cfg=dict(type='SyncBN'),
        act_cfg=dict(type='SiLU', inplace=True)),
    bbox_head=dict(
        type='RTMDetInsSepBNHead',
        num_classes=1,
        in_channels=256,
        stacked_convs=2,
        share_conv=True,
        pred_kernel_size=1,
        feat_channels=256,
        act_cfg=dict(type='SiLU', inplace=True),
        norm_cfg=dict(type='SyncBN', requires_grad=True),
        anchor_generator=dict(
            type='MlvlPointGenerator', offset=0, strides=[
                8,
                16,
                32,
            ]),
        bbox_coder=dict(type='DistancePointBBoxCoder'),
        loss_cls=dict(
            type='QualityFocalLoss',
            use_sigmoid=True,
            beta=2.0,
            loss_weight=1.0),
        loss_bbox=dict(type='GIoULoss', loss_weight=2.0),
        loss_mask=dict(
            type='DiceLoss', loss_weight=2.0, eps=5e-06, reduction='mean')),
    train_cfg=dict(
        assigner=dict(type='DynamicSoftLabelAssigner', topk=13),
        allowed_border=-1,
        pos_weight=-1,
        debug=False),
    test_cfg=dict(
        nms_pre=1000,
        min_bbox_size=0,
        score_thr=0.05,
        nms=dict(type='nms', iou_threshold=0.6),
        max_per_img=100,
        mask_thr_binary=0.5))
train_pipeline_stage2 = [
    dict(type='LoadImageFromFile', backend_args=None),
    dict(
        type='LoadAnnotations',
        with_bbox=True,
        with_mask=True,
        poly2mask=False),
    dict(
        type='RandomResize',
        scale=(
            640,
            640,
        ),
        ratio_range=(
            0.1,
            2.0,
        ),
        keep_ratio=True),
    dict(
        type='RandomCrop',
        crop_size=(
            640,
            640,
        ),
        recompute_bbox=True,
        allow_negative_crop=True),
    dict(type='FilterAnnotations', min_gt_bbox_wh=(
        1,
        1,
    )),
    dict(type='YOLOXHSVRandomAug'),
    dict(type='RandomFlip', prob=0.5),
    dict(type='Pad', size=(
        640,
        640,
    ), pad_val=dict(img=(
        114,
        114,
        114,
    ))),
    dict(type='PackDetInputs'),
]
metainfo = dict(
    classes='[redacted]', palette=[
        (
            220,
            20,
            60,
        ),
    ])
max_epochs = 300
stage2_num_epochs = 20
base_lr = 0.004
interval = 10
custom_hooks = [
    dict(
        type='EMAHook',
        ema_type='ExpMomentumEMA',
        momentum=0.0002,
        update_buffers=True,
        priority=49),
    dict(
        type='PipelineSwitchHook',
        switch_epoch=280,
        switch_pipeline=[
            dict(type='LoadImageFromFile', backend_args=None),
            dict(
                type='LoadAnnotations',
                with_bbox=True,
                with_mask=True,
                poly2mask=False),
            dict(
                type='RandomResize',
                scale=(
                    640,
                    640,
                ),
                ratio_range=(
                    0.1,
                    2.0,
                ),
                keep_ratio=True),
            dict(
                type='RandomCrop',
                crop_size=(
                    640,
                    640,
                ),
                recompute_bbox=True,
                allow_negative_crop=True),
            dict(type='FilterAnnotations', min_gt_bbox_wh=(
                1,
                1,
            )),
            dict(type='YOLOXHSVRandomAug'),
            dict(type='RandomFlip', prob=0.5),
            dict(
                type='Pad',
                size=(
                    640,
                    640,
                ),
                pad_val=dict(img=(
                    114,
                    114,
                    114,
                ))),
            dict(type='PackDetInputs'),
        ]),
]
launcher = 'none'
work_dir = '/workspace/models/rtmdet_ins_l'
```

**Comment by roman-duris** [comment]
Hi I found a way to significantly improve the segmentation precision for RTMDet ins models. It seems that it does not have a significant speed penalty either. 
In the RTMDet ins head change the ```mask_loss_stride``` to some lower number, for me value of ```1``` worked fine.
Also weight of the  mask loss can be increased. Of course this will depend on your dataset. 
 
This of course increases duration and memory used during training significantly but during the inference everything seems as fast as before. 
Precision Before the change:
```{"coco/bbox_mAP": 0.755, "coco/bbox_mAP_50": 0.833, "coco/bbox_mAP_75": 0.789, "coco/bbox_mAP_s": 0.485, "coco/bbox_mAP_m": 0.955, "coco/bbox_mAP_l": 1.0, "coco/segm_mAP": 0.609, "coco/segm_mAP_50": 0.789, "coco/segm_mAP_75": 0.739, "coco/segm_mAP_s": 0.315, "coco/segm_mAP_m": 0.82, "coco/segm_mAP_l": 0.968, "data_time": 0.00667271894567153, "time": 0.3168791181900922, "step": 300}```
Precision After the change:
```{"coco/bbox_mAP": 0.761, "coco/bbox_mAP_50": 0.835, "coco/bbox_mAP_75": 0.788, "coco/bbox_mAP_s": 0.498, "coco/bbox_mAP_m": 0.956, "coco/bbox_mAP_l": 0.989, "coco/segm_mAP": 0.712, "coco/segm_mAP_50": 0.822, "coco/segm_mAP_75": 0.776, "coco/segm_mAP_s": 0.418, "coco/segm_mAP_m": 0.927, "coco/segm_mAP_l": 0.991, "data_time": 0.0025061190128326414, "time": 0.13585618734359742, "step": 300}```

One can note that the small particle precision of segmentation has increased by   0.1 AP and the overall segmentation precision has also increased by more than 0.1 AP.
